universal-login-relayer/test/utils/utils.js

```diff
@@ -22,12 +24,45 @@ describe('Tools test', async () => {
     [wallet] = await getWallets(provider);
   });
 
-  it('Should return correct message signature', async () => {
-    const from = wallet.address;
-    const signature = await messageSignature(wallet, wallet.address, from, value, data, nonce, gasToken, gasPrice, gasLimit);
-    const message = utils.arrayify(utils.solidityKeccak256(
-      ['address', 'address', 'uint256', 'bytes', 'uint256', 'address', 'uint', 'uint'],
-      [wallet.address, from, value, data, nonce, gasToken, gasPrice, gasLimit]));
-    expect(Wallet.verifyMessage(message, signature)).to.eq(wallet.address);
+  describe('messageSignature', async () => {
+    it('Should return correct message signature', async () => {
+      const from = wallet.address;
+      const signature = await messageSignature(wallet, wallet.address, from, value, data, nonce, gasToken, gasPrice, gasLimit);
+      const message = utils.arrayify(utils.solidityKeccak256(
+        ['address', 'address', 'uint256', 'bytes', 'uint256', 'address', 'uint', 'uint'],
+        [wallet.address, from, value, data, nonce, gasToken, gasPrice, gasLimit]));
+      expect(Wallet.verifyMessage(message, signature)).to.eq(wallet.address);
+    });  
+  });
+
+  describe('isEnoughGasLimit', async () => {
+    it('Should return true if is enough gasLimit', async () => {
+      expect(isEnoughGasLimit(gasLimit - 1 , gasLimit)).to.be.true;
+      expect(isEnoughGasLimit(gasLimit, gasLimit)).to.be.true;
+    });
+
+    it('Should return false if is not enough gasLimit', async () => {
+      expect(isEnoughGasLimit(gasLimit * 2, gasLimit)).to.be.false;
+      expect(isEnoughGasLimit(gasLimit + 1, gasLimit)).to.be.false;
+    });
+  });
+
+  describe('hasEnoughToken', async () => {
+    let token;
+    let identity;
+
+    before(async () => {
+      token = await deployContract(wallet, MockToken, []);
+      identity = await deployContract(wallet, ERC725ApprovalScheme, [addressToBytes32(wallet.address)]);
+    });
+
+    it('Should return true if contract has enough token', async () => {
+      await token.transfer(identity.address, utils.parseEther('1'));
+      expect(await hasEnoughToken(token.address, identity.address, gasLimit, wallet)).to.be.true;
+    });
+
+    it('Should return false if contract has not enough token', async () => {
+      expect(await hasEnoughToken(token.address, identity.address, utils.parseEther('2'), wallet)).to.be.false;
+    });
```

Comment:
We should implement more tests, this is a crucial piece for security. E.g.:
- what if passed address is not a for a token address?
- what if passed address is ether?
- a couple expects for different values as a sanity checks

Reply:
✅